src/core/port-range-access.c

```diff
@@ -0,0 +1,295 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+/***
+  This file is part of systemd.
+
+  Copyright 2016 Daniel Mack
+  Copyright 2017 Intel Corporation.
+
+  systemd is free software; you can redistribute it and/or modify it
+  under the terms of the GNU Lesser General Public License as published by
+  the Free Software Foundation; either version 2.1 of the License, or
+  (at your option) any later version.
+
+  systemd is distributed in the hope that it will be useful, but
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include <stdio.h>
+#include <stdlib.h>
+
+#include "alloc-util.h"
+#include "bpf-firewall.h"
+#include "extract-word.h"
+#include "port-range-access.h"
+#include "parse-util.h"
+#include "string-util.h"
+#include "socket-protocol-list.h"
+
+static int create_series(
+                uint16_t range_start,
+                uint16_t range_end,
+                unsigned char protocol,
+                PortRangeAccessItem **list) {
+        uint biggest_exp = 1, biggest_block = 0;
+        uint numbers = range_end - range_start + 1;
+        uint next_candidate, best_candidate;
+        uint16_t mask;
+        PortRangeAccessItem *a;
+        int i, r;
+
+        /*
+         * We want to use prefix matching to allow for fast port range
+         * processing in kernel. This function finds the biggest prefix that is
+         * fully contained within the port range. It then calls itself
+         * recursively for the non-covered parts of the range.
+         *
+         * For example typical ephemeral ports:
+         *   32768-61000 -> [ 32768/2 49152/3 57344/5 59392/6 60416/7 60928/10 60992/13 61000/16 ]
+         */
+
+        a = new0(PortRangeAccessItem, 1);
+        if (!a)
+                return -ENOMEM;
+
+        a->protocol = protocol;
+
+        if (range_start == range_end) {
+                a->prefixlen = 16;
+                a->port = range_start;
+                LIST_APPEND(items, *list, a);
+                return 0;
+        }
+
+        /*
+         * Variable numbers is the inclusive range length. Count how many
+         * potential prefixes we could fit in the inclusive range. Biggest
+         * exponent will be at least 2.
+         */
+
+        while (biggest_block <= numbers)
+                biggest_block = 1 << biggest_exp++;
+
+        /* Find the largest block that will fit (in prefix sense) into the range. */
+        for (i = biggest_exp-2; i >= 0; i--) {
+
+                uint16_t block = 1 << i;
```

Comment:
you should really use 1U here I figure, otherwise you generate a signed value here that you then assigned to an unsigned variable. This is problematic as << sign-extends on C.